src/os/bluestore/BlueFS.cc

```diff
@@ -2229,6 +2258,177 @@ int64_t BlueFS::_read_random(
   return ret;
 }
 
+void BlueFS::_wal_update_size(FileRef file, uint64_t flush_offset, uint64_t increment) {
+  dout(20) 
+      << fmt::format(
+        "{} updating WAL file {} for range {:#x}~{:#x} limit is {:#x}", 
+        __func__, file->fnode.ino, flush_offset, increment, file->fnode.wal_limit) 
+      << dendl;
+  ceph_assert(file->is_wal_read_loaded);
+  ceph_assert((flush_offset == 0 && file->wal_flushes.empty()) || (flush_offset == file->wal_flushes.back().end_offset()));
+
+  FileReader *h = new FileReader(file, cct->_conf->bluefs_max_prefetch, false, true);
+
+  uint64_t flush_end = flush_offset + increment;
+  while (flush_offset < file->fnode.wal_limit) {
+    // read first part of wal flush
+    bufferlist bl;
+    uint64_t read_result = (uint64_t)_read(h, flush_offset, sizeof(File::WALFlush::WALLength), &bl, nullptr);
+    if (read_result < sizeof(File::WALFlush::WALLength)) {
+      dout(20) << fmt::format("{} cannot read flush length, most likely we are out of bounds. flush_offset={:#X}", __func__, flush_offset) << dendl;
+      break;
+    }
+    File::WALFlush::WALLength flush_length_le(0);
+    flush_length_le = *((File::WALFlush::WALLength*)bl.c_str());
+    uint64_t flush_length = flush_length_le;
+    dout(20) << __func__ << " flush_length " << flush_length << dendl;
+
+    // read marker
+    bl.clear();
+    uint64_t marker_offset = flush_offset+sizeof(File::WALFlush::WALLength)+flush_length;
+    read_result = _read(h, marker_offset, sizeof(File::WALFlush::WALMarker), &bl, nullptr);
+    File::WALFlush::WALMarker marker_le(0);
+    if (read_result < sizeof(File::WALFlush::WALMarker)) {
+      dout(20) << fmt::format("{} cannot read marker, most likely we are out of bounds. flush_offset={:#X}, marker_offset={:#X}", __func__, flush_offset, marker_offset) << dendl;
+      break;
+    }
+    marker_le = *((File::WALFlush::WALMarker*)bl.c_str());
+    uint64_t marker = marker_le;
+    if (marker != file->fnode.ino) {
+      // EOF or corruption
+      break;
+    }
+
+    uint64_t increase = flush_length+(sizeof(File::WALFlush::WALLength)+sizeof(File::WALFlush::WALMarker));
+    dout(20) << fmt::format("{} adding flush {:#x}~{:#x}", __func__, flush_offset, flush_length) << dendl;
+    file->wal_flushes.push_back({flush_offset, flush_length});
```

Comment:
flush_offset is an offset of the envelop header, flush_length - actual envelop payload length here, right?
IMO that's a bit confusing - you better preserve payload offset here as well. I presume one doesn't need envelop header beyond this point...

Reply:
flush_offset and flush_length is the minimum amount of information needed to extrapolate all other information. I think I can reuse the get_payload_offset for comment below.